### PR TITLE
Create ThingReference class (WIP)

### PR DESCRIPTION
**Pull request by cdrini**
In Python 3, we can't override the __class__ object inside
Thing._getdata (not entirely sure why). As an alternative, try to
separate the cases where a Thing is just a shallow reference without
data and when it's been populated with information from the database.

This causes loads of things to breaks, and needs way more testing.

### CODE REVIEW

File: infogami/infobase/client.py [hunk 1]
```diff
@@ -691,44 +691,58 @@ def __repr__(self): return ""
...
+                    type = type.key
+                elif isinstance(type, dict):
+                    type = type['key']
+
+                # just to be safe
+                if not isinstance(type, string_types):
+                    type = None
+        except Exception as e:
+            # just for extra safety
+            print('ERROR:', str(e), file=web.debug)
```
Comment (cclauss) [inline infogami/infobase/client.py:717]:
```suggestion
            traceback.print_exc()
```

Comment (cclauss) [inline infogami/infobase/client.py:717]:
```suggestion
        except Exception:
            # just for extra safety
            traceback.print_exc()
```

File: infogami/infobase/client.py [hunk 2]
```diff
@@ -691,44 +691,58 @@ def __repr__(self): return ""
...
+            type = None
+        klass = _thing_class_registry.get(type) or _thing_class_registry.get(None)
     return klass(site, key, data, revision)
 
-class Thing:
+
+class BaseThing:
```
Comment (cclauss) [inline infogami/infobase/client.py:723]:
```suggestion
class BaseThing(object):
```

**Comment by cclauss** [comment]
@cdrini Status of this one?

**Comment by cdrini** [comment]
I think we found another fix for this; closing.

**Pull request closed without merging**